src/modules/complianceengine/src/lib/procedures/EnsureShadowContains.cpp

```diff
@@ -0,0 +1,443 @@
+        }
+        default:
+            break;
+    }
+
+    auto intValue = AsInt(value);
+    if (!intValue.HasValue())
+    {
+        return Error("Invalid integer value: " + value, EINVAL);
+    }
+
+    switch (field)
+    {
+        case Field::LastChange:
+            return IntegerComparison(entry.sp_lstchg, intValue.Value(), operation);
+        case Field::MinAge:
+            return IntegerComparison(entry.sp_min, intValue.Value(), operation);
+        case Field::MaxAge:
+            return IntegerComparison(entry.sp_max, intValue.Value(), operation);
+        case Field::WarnPeriod:
+            return IntegerComparison(entry.sp_warn, intValue.Value(), operation);
+        case Field::InactivityPeriod:
+            return IntegerComparison(entry.sp_inact, intValue.Value(), operation);
+        case Field::ExpirationDate:
+            return IntegerComparison(entry.sp_expire, intValue.Value(), operation);
+        default:
+            break;
+    }
+
+    return Error(PrettyFieldName(field) + " field comparison is not supported", EINVAL);
+}
+} // anonymous namespace
+
+namespace ComplianceEngine
+{
+AUDIT_FN(EnsureShadowContains, "username:A pattern or value to match usernames against",
+    "username_operation:A comparison operation for the username parameter::(eq|ne|lt|le|gt|ge|match)",
+    "field:The /etc/shadow entry field to match "
+    "against:M:(password|last_change|min_age|max_age|warn_period|inactivity_period|expiration_date|encryption_method)",
+    "value:A pattern or value to match against the specified field:M",
+    "operation:A comparison operation for the value parameter:M:(eq|ne|lt|le|gt|ge|match)")
+{
+    UNUSED(context);
+
+    Optional<string> username;
+    auto it = args.find("username");
+    if (it != args.end())
+    {
+        username = std::move(it->second);
+    }
+
+    auto usernameOperation = Operation::Equal;
+    it = args.find("username_operation");
+    if (it != args.end())
+    {
+        auto operation = ParseOperation(it->second);
+        if (!operation.HasValue())
+        {
+            return operation.Error();
+        }
+        usernameOperation = operation.Value();
+    }
+
+    it = args.find("field");
+    if (it == args.end())
+    {
+        return Error("Missing 'field' parameter", EINVAL);
+    }
+    auto field = ParseFieldName(it->second);
+    if (!field.HasValue())
+    {
+        return field.Error();
+    }
+
+    it = args.find("value");
+    if (it == args.end())
+    {
+        return Error("Missing 'value' parameter", EINVAL);
+    }
+    auto value = std::move(it->second);
+
+    it = args.find("operation");
+    if (it == args.end())
+    {
+        return Error("Missing 'operation' parameter", EINVAL);
+    }
+    auto operation = ParseOperation(it->second);
+    if (!operation.HasValue())
+    {
+        return operation.Error();
+    }
+
+    assert(field.HasValue());
+    assert(operation.HasValue());
+
+    // Iterate over all users
+    setspent();
```

Comment:
Could we use the reentrant version, and maybe allow for test-provided filename for shadow?